### ISSUE TITLE
calling methods on Array constructor (Array.filter, etc.)

### ISSUE THREAD

**Issue by oberhamsi** [mozilla/rhino#213]
there's a couple of places in our code were we have something like the following:

    var args = Array.filter(arguments, function(p) p != "")

the above used to return the filtered Array but the behaviour changed to returning `undefined` with a recent rhino commit https://github.com/mozilla/rhino/commit/5fda169c76e6b1a569b9c7b715b0af6345b84fde

my firefox behaves like the old rhino whereas chrome doesn't even have `Array.filter` defined.  i'm certainly more used to seeing it done less elegantly: `Array.prototype.filter.call`

any idea what's going on there?

**Comment by eshepelyuk** [comment]
Array constructor has no such methods in [ES 5.1 spec] (http://www.ecma-international.org/ecma-262/5.1/#sec-15.4.3). And even in [older specs](http://www.ecma-international.org/publications/standards/Ecma-262-arch.htm).
So proper way of using it is `Array.prototype.filter.call` or `Array.prototype.filter.apply`.

**Comment by oberhamsi** [comment]
thanks the quick fact check. we'll fix our code.

not sure if you consider this a bug then:

    Rhino 1.7.8-SNAPSHOT 2015 07 01
    js> Array.filter
    function filter() { [native code for Array.filter, arity=1] }

**Comment by sainaen** [comment]
@eshepelyuk 
Yes, but it works right now, just has a bug in it: now we don't pass correct Id to `iterativeMethod` when it's a constructor property.

**Comment by eshepelyuk** [comment]
Apparently, spec violation. As for me - should be fixed to work according to spec.
I think the same pattern will apply for other classes, not just `Array`.

@gbrail what do you think about it ?

**Comment by sainaen** [comment]
The same thing with NativeString. NativeDate and NativeObject are fine in that regard.

**Comment by gbrail** [comment]
Well, "Array.filter" (on the constructor) isn't in the spec. But it's in
Rhino, and it's been in there for a very long time. I think that the safest
thing to do is to make it work the way it used to work if we can without
breaking ES6 compatibility for other cases.

I'm not sure why this and other constructor functions are present in Rhino.

For instance, the Mozilla (non-canonical) JavaScript docs only mention
"from," "of," "observe," and "isArray." Rhino, however, also defines a
whole bunch more. (However, it does not yet define "from" and "of", which
are in ES6.)

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array

I am certain that if Simon's project is depending on this old behavior,
then others will be as well. (I'm also sad that apparently we never had a
test for this.)

I think that the safest course is to prevent breaking old code and make the
old behavior work again even though these functions are not in the spec.

On Wed, Jul 1, 2015 at 11:52 AM, Evgeny Shepelyuk <notifications@github.com>
wrote:

> Apparently, spec violation. As for me - should be fixed to work according
> to spec.
> I think the same pattern will apply for other classes, not just Array.
>
> @gbrail <https://github.com/gbrail> what do you think about it ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla/rhino/issues/213#issuecomment-117794299>.
>



-- 
*Greg Brail* | *apigee <https://apigee.com/>* | twitter @gbrail
<http://twitter.com/gbrail> @apigee <https://twitter.com/apigee>
<http://iloveapis.com/>

**Comment by oberhamsi** [comment]
reopening per discussion
